### PR TITLE
Modify the CalendarItem logic to better account for all day events

### PR DESCRIPTION
**Pull request by annielcook**
This diff better handles CalendarItem (aka events) which are all day events. Previously, the start and end fields on `CalendarItem` were `DateTimeField`s which meant that they had time and timezone info. This messes with all day events because instead of associating the all day event with a specific day, Exchange now tries to fit the event into the proper time slot (like 5pm 12/2 - 5pm 12/3). On top of all of this, exchange has logic that for all day events that don't start or end at midnight, it should adjust the start to the previous midnight and the end to the following midnight. This means that these all day events with times will now be shifted to span even more days. 

This diff fixes that by creating a `GenericDateField` that both start and end use. During sync is when we call `from_xml`. For the new class, this method checks if its an all day event and if so, uses logic copied from `DateField` otherwise uses logic from `DateTimeField`. For syncback we can just use the generic `to_xml` function defined on the parent Field.  The trick for syncback is avoiding the check of `if not isinstance(value, self.value_cls):`. The new generic field class does not have a `value_cls` since it can be either `EWSDate` or `EWSDateTime` depending on if its all day or not, so this would error. To avoid this, I return early. 

The final change in this diff is not adding timezones if the event is all day. This diff relies on a change in our syncback code where we conditionally build `EWSDate` and `EWSDateTime` objects for the start and end. 

This code is kind of messy but thats mostly to get it to work with the convoluted way in which exchangelib is set up. I tested this for sync and syncback and both worked. I would suggest you test as well! You can `pip uninstall exchangelib` and then install this branch. Locally I ran `pip install git+file:///Users/anniecook/code/Nylas/exchangelib@all-day-event-handling`

### CODE REVIEW

File: exchangelib/fields.py [hunk 1]
```diff
@@ -529,31 +531,44 @@ def to_xml(self, value, version):
         return set_xml_value(field_elem, base64.b64encode(value).decode('ascii'), version=version)
 
 
-class DateField(FieldURIField):
-    value_cls = EWSDate
-
+class GenericDateField(FieldURIField):
```
Comment (betsyalegria) [inline exchangelib/fields.py:534]: a docstring for this class might be helpful to add!

File: exchangelib/fields.py [hunk 2]
```diff
@@ -529,31 +531,44 @@ def to_xml(self, value, version):
...
     def from_xml(self, elem, account):
         val = self._get_val_from_elem(elem)
+        is_all_day = elem.find('{http://schemas.microsoft.com/exchange/services/2006/types}IsAllDayEvent')
```
Comment (betsyalegria) [inline exchangelib/fields.py]: interesting, is there a file that stores these urls or something? it feels a bit out of place but if not, then I'm cool with keeping it here!

**Review by betsyalegria — APPROVED** [review]
A few nits, but overall I think this looks good

File: exchangelib/fields.py (hunk 2 above)
Comment (annielcook) [reply exchangelib/fields.py]: Oh ya this was a total hack. There is no file that has these in their entirety, instead I followed the logic for how the fields are fetched, which happens by calling the `response_tag` function on instance of the `FieldUriField` class and dynamically generating '{%s}%s' % (TNS, self.field_uri_postfix) where TNS is the http part and the field_uri_postfix is what comes after the `:` in the field_uri param to the field  `BooleanField('is_all_day', field_uri='calendar:IsAllDayEvent', is_required=True, default=False)` for example. 

Normally a field is creating its own response tag. In this code, however, I want the GenericDateField to create to find a different field (is_all_day) and thus it has to generate the response tag this way. 

I can clean this up by referencing TNS.

File: exchangelib/fields.py (hunk 1 above)
Comment (annielcook) [reply exchangelib/fields.py:534]: Good idea

**Pull request merged**